scripts/manpages/man1/bm.1
.TH BM 1 "20 April 2016" "version 1.3"

.SH NAME
bm - Bookmarks for file-system locations

.SH SYNOPSIS
.BR "bm" " [" "-hl" "] [" "--cleartmp" "] [" "-s "
.IR "name"
.RB "[" "-t" "]] [" "-g "
.IR "name" "]"
.RB "[" "-l" "] [" "-r "
.IR "name" "]"
.RB "[" "-p "
.IR "name" "]"

.SH DESCRIPTION
.B bm
allows to bookmark locations file-system locations.  This makes it possible to
quickly navigate to the bookmarked locations by using their bookmark name
instead of the full path.

.PP
Bookmarks are grouped into temporary and permanent bookmarks.  All temperorary
bookmarks can be deleted by using the
.B --cleartmp
option.  Permament bookmarks have to be removed individually by using the
.B --remove
option.

.SH OPTIONS

.IP "-h, --help"
Displays this help message and exits.

.IP "-s, --set"
The current working directory is saved as a bookmark.  If a
.I name
is provided, its used as the bookmark name.  Otherwise, the user is prompted to
enter the name of the new bookmark.  Existing bookmarks are overwritten without
warning.  If the
.B -t
switch is active, the bookmark is stored as a temporary bookmark, otherwise it
is stored permanently.

.IP "-g, --goto"
The current working directory is changed to the location, associated with the
specified
.IR "name" "."
If no name is provided, the user is prompted to specify the name interactively.

.IP "-l, --list"
Lists all permanent and temporary bookmarks.  Broken links are colored red.

.IP "-r, --remove"
Removes the bookmark with the specified
.IR "name" "."
If no name is provided, the user is prompted to specify the name interactively.

.IP "-p, --print"
Prints the path of the bookmarked location with the specified
.IR "name" "."
If no name is provided, the user is prompted to specify the name interactively.

.IP "-t, --tmp"
Influences the
.B -s
option.  If the
.B -t
switch is used, the bookmark is stored as a temporary bookmark, otherwise it is
stored as a permanent bookmark.

.IP "-ct, --cleartmp"
Deletes all temporary bookmarks.

.SH AUTHORS
Wolfram Reinke <wolframreinke@web.de>


scripts/bookmarks/bm.sh
#! /bin/bash

# ==============================================================================
#
# This script manages bookmark for the user. He can add any location on the
# computer to his bookmarks, and jump to this bookmark by specifying its name.
# Also, the user can search the list of stored bookmarks
#
# ==============================================================================

DEBUG="false"

function _debug
{
    if [ ${DEBUG} == "true" ]
    then
        echo -e "[DEBUG]\t${1}"
    fi

    return 0
}

function _display_help
{
    cat ${HOME}/scripts/bookmarks/bm.help
    return 0
}

function _bm_main
{
    local BOOKMARK_STORE="${HOME}/scripts/bookmarks"
    local BOOKMARK_STORE_PER="${BOOKMARK_STORE}/permanent"
    local BOOKMARK_STORE_TMP="${HOME}/scripts/bookmarks/temporary"
    local BOOKMARK_SCRIPT="${HOME}/scripts/bookmarks/bm.sh"

    # --------------------------------
    # Parse the command line arguments
    # --------------------------------

    # These values are set during the parsing procedure
    local ARGUMENT="undefined"
    local ARG_PATH="undefined"
    local MODE="default"
    local TEMPORARY="false"

    # iterate over all given arguments
    while [ $# -gt 0 ]
    do
        CURRENT_OPT=${1}
        shift

        case ${CURRENT_OPT} in

        # Display the help message
        -h|--help)
            _display_help;
            return 0
            ;;

        # switch "save" mode. The current position will be saved with
        # the given name
        -s|--set)
            MODE="save"
            ;;

        # switch to "goto" mode. The current position will be changed to
        # the specified bookmark
        -g|--goto)
            MODE="goto"
            ;;

        # display all available bookmarks in aplhabetical order
        -l|--list)

            for DIRECTORY in permanent temporary; do

                _debug "${DIRECTORY}"

                echo -e "\n${DIRECTORY}" | tr '[:lower:]' '[:upper:]'
                printf "%80s\n" | tr ' ' -
                for FILENAME in ${BOOKMARK_STORE}/${DIRECTORY}/*.bm; do

                    if [ "${FILENAME}" != "${BOOKMARK_STORE}/${DIRECTORY}/*.bm" ]; then

                        _debug "${FILENAME}"

                        local LEFT=$(echo ${FILENAME} | grep -Poi '(?<=\/)((?!\/).)*?(?=\.bm)')
                        local RIGHT=$(cat ${FILENAME})
                        if [ ! -e "${RIGHT}" ]; then
                            printf "%-20s - \e[0;31m%s\e[0m\n" ${LEFT} ${RIGHT}
                        else
                            printf "%-20s - %s\n" ${LEFT} ${RIGHT}
                        fi
                    else
                        echo "(none)"
                    fi
                done
                echo ""
            done

            return 0;
            ;;

        # delete the specified bookmark
        -r|--remove)
            MODE="rm"
            ;;

                -p|--print)
                        MODE="print"
                        ;;

        -t|--tmp)
            TEMPORARY="true"
            ;;

        -ct|--cleartmp)
            rm ${BOOKMARK_STORE_TMP}/*.bm > /dev/null 2>&1;
            return 0
            ;;

        # unknown options start with a -, but the rest is unrecognized
       -*)
            echo "Unknown option ${CURRENT_OPT}"
            return 1
            ;;

        # the argument to the chosen operation
        *)
            ARGUMENT=$(echo ${CURRENT_OPT} | sed 's_^\([^/]*\).*_\1_g')
            ARGUMENT=${CURRENT_OPT%%/*}
            ARG_PATH=${CURRENT_OPT#*/}

            if [ "${ARGUMENT}" == "$ARG_PATH" ]; then
                    ARG_PATH="."
            fi
            ;;

        esac
    done

    # the user has not chosen either --list --save or --goto.  if the specified
    # bookmark exists, the corresponding directory is opened, if not, a new
    # bookmark with the specified name is created at the current position
    if [ ${MODE} == "default" ]; then

        if [ ${ARGUMENT} == "undefined" ]; then
            echo -n "Name of the bookmark: "
            read ARGUMENT
        fi

        if [   -e ${BOOKMARK_STORE_PER}/${ARGUMENT}".bm" \
            -o -e ${BOOKMARK_STORE_TMP}/${ARGUMENT}".bm" ]; then
            _bm_main --goto $ARGUMENT/$ARG_PATH
        else

            if [ ${TEMPORARY} == "false" ]; then
                _bm_main --set ${ARGUMENT}
            else
                _bm_main --tmp --set ${ARGUMENT}
            fi
        fi

    elif [ ${MODE} == "save" ]; then

        if [ ${ARGUMENT} == "undefined" ]; then

            echo -n "Name for the new bookmark: "
            read ARGUMENT
        else

            echo -n "Are you sure, you want to create a new bookmark? (Y/N): "
            read RESPONSE

            if [ "${RESPONSE}" != "Y" ]; then
                echo Aborting.
                return
            fi
        fi

        if [ ${TEMPORARY} == "false" ]; then
            echo ${PWD} > ${BOOKMARK_STORE_PER}/${ARGUMENT}".bm"
        else
            echo ${PWD} > ${BOOKMARK_STORE_TMP}/${ARGUMENT}".bm"
        fi

    elif [ ${MODE} == "rm" ]; then

        if [ ${ARGUMENT} == "undefined" ]; then
            echo -n "Name of the bookmark to delete: "
            read ARGUMENT
        fi


        rm ${BOOKMARK_STORE_PER}/${ARGUMENT}".bm" > /dev/null 2>&1
        rm ${BOOKMARK_STORE_TMP}/${ARGUMENT}".bm" > /dev/null 2>&1

        return 0
        elif [ ${MODE} == "print" ]; then

                if [ ${ARGUMENT} == "undefined" ]; then
                        echo -n "Name of the bookmark to primt: "
                        read ARGUMENT
                fi

                TARGET=$(cat ${BOOKMARK_STORE_PER}/${ARGUMENT}".bm" 2> /dev/null)
                if [ -z ${TARGET} ]; then
                        TARGET=$(cat ${BOOKMARK_STORE_TMP}/${ARGUMENT}".bm" 2> /dev/null)
                fi

                echo ${TARGET}
    else
        if [ ${ARGUMENT} == "undefined" ]; then
            echo -n "Name of the desired bookmark: "
            read ARGUMENT
        fi

        _debug "ARGUMENT = ${ARGUMENT}"

        TARGET=$(cat ${BOOKMARK_STORE_PER}/${ARGUMENT}".bm" 2> /dev/null)

        _debug "TARGET = $TARGET"

        if [ -z ${TARGET} ]; then
            TARGET=$(cat ${BOOKMARK_STORE_TMP}/${ARGUMENT}".bm" 2> /dev/null)
        fi

        _debug "TARGET = ${TARGET}"

        if [ ${TARGET} ]; then
            cd ${TARGET}/${ARG_PATH}
        else
            echo "No such bookmark."
            return 1
        fi
    fi
}

alias bm=_bm_main
